### PR TITLE
feat(autoware_auto_perception_rviz_plugin): add perception objects pointcloud publisher

### PR DESCRIPTION
**Pull request by lexavtanke**
## Description

This PR adding new function to autoware_auto_perception_rviz_plugin. Now it has ability to publish point cloud of detected objects. It improve visualization of perception pipeline. 
This feature can be enabled/disabled in control panel of the plugin by toggling Publish Objects Pointcloud property.  
New point cloud plugin should be added to autoware config to visualize output of the plugin. Like it's shown in this [video](https://youtu.be/cRTaotGZ1x4). 

Output topics for point clouds:

- /rviz2/output/predicted_objects_pointcloud
- /rviz2/output/detected_objects_pointcloud
- /rviz2/output/tracked_objects_pointcloud 

[Video demo](https://youtu.be/lJhmTXtJ7jY) with predicted objects from autoware demo bagfile. 

## Related links

This feature was discussed here [2860](https://github.com/autowarefoundation/autoware.universe/issues/2860).
## Tests performed

This feature was tested manually against sample rosbag from Rosbag replay simulation demo. I didn't find any reduces in performance. 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by lexavtanke** [comment]
#2860

**Comment by kaancolak** [comment]
Hi Alexey, it works as expected but I have some comments. 

Maybe instead of using another topic for the object point cloud, we can solve it under the same display plugin. 

As an alternative solution could be to collect specific sizes of the point cloud, like queue structure then into the addMarker function if the "Publish Object Pointcloud" checkbox is clicked, we can add points as a marker like the other 3D bounding boxes. 3D objects already generated "/perception/obstackle_segmentatin/pointcloud", probably there won't be any timing issue.

**Review by kaancolak — COMMENTED** [review]
Hi Alexey @lexavtanke, I left out some comments.

File: common/autoware_auto_perception_rviz_plugin/src/object_detection/predicted_objects_display.cpp [hunk 1]
```diff
@@ -210,6 +211,121 @@ void PredictedObjectsDisplay::update(float wall_dt, float ros_dt)
...
+    return;
+  }
+
+  objs_buffer.clear();
+  for (const auto & object : transformed_objects.objects) {
+    std::vector<autoware_auto_perception_msgs::msg::ObjectClassification> labels =
+      object.classification;
+    object_info info = {
+      object.shape, object.kinematics.initial_pose_with_covariance.pose, object.classification};
+    objs_buffer.push_back(info);
```
[the remark this answers is not in the document]
Comment (lexavtanke) [reply common/autoware_auto_perception_rviz_plugin/src/object_detection/predicted_objects_display.cpp]: There is no objs_buffer but I still use conversion to object_info to unify data for filterPolygon function and keep in base class.

File: common/autoware_auto_perception_rviz_plugin/src/object_detection/predicted_objects_display.cpp [hunk 2]
```diff
@@ -210,6 +211,121 @@ void PredictedObjectsDisplay::update(float wall_dt, float ros_dt)
...
 
+void PredictedObjectsDisplay::onInitialize()
+{
+  ObjectPolygonDisplayBase::onInitialize();
+  // get access to rviz node to sub and to pub to topics
+  rclcpp::Node::SharedPtr raw_node = this->context_->getRosNodeAbstraction().lock()->get_raw_node();
+
+  sync_ptr = std::make_shared<Sync>(
+    SyncPolicy(10), perception_objects_subscription, pointcloud_subscription);
+  sync_ptr->registerCallback(&PredictedObjectsDisplay::onObjectsAndObstaclePointCloud, this);
```
[the remark this answers is not in the document]
Comment (lexavtanke) [reply common/autoware_auto_perception_rviz_plugin/src/object_detection/predicted_objects_display.cpp]: I've updated this part. Now there is separate callback for poincloud messages.

File: common/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp [hunk 3]
```diff
@@ -147,6 +226,62 @@ class AUTOWARE_AUTO_PERCEPTION_RVIZ_PLUGIN_PUBLIC ObjectPolygonDisplayBase
     m_marker_common.addMessage(markers_ptr);
   }
 
+  void add_pointcloud(const sensor_msgs::msg::PointCloud2::ConstSharedPtr cloud)
```
[the remark this answers is not in the document]
Comment (lexavtanke) [reply common/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp:309]: It is already build-in.

**Review by Shin-kyoto — CHANGES_REQUESTED** [review]
@lexavtanke 
Can you resolve conflicts?

**Comment by Shin-kyoto** [comment]
@lexavtanke 

Thank you for creating the PR.
Would it be possible for you to resolve the conflicts? 

To keep Autoware healthy, we close PRs if there is no response for more than a week after requesting conflict resolution. If this PR is still active, please let us know. Even if this PR is closed, you are welcome to reopen it once the conflicts have been resolved.

**Comment by Shin-kyoto** [comment]
@lexavtanke 

friendly ping

**Comment by Shin-kyoto** [comment]
@lexavtanke

friendly ping

**Comment by lexavtanke** [comment]
Hi @Shin-kyoto 
Sorry for the late reply.

As I know there are PRs from @KhalilSelyan https://github.com/autowarefoundation/autoware.universe/pull/7095 and https://github.com/autowarefoundation/autoware.universe/pull/8247.
He continues my work on this task. And I think this one PR could be closed.

**Comment by Shin-kyoto** [comment]
@lexavtanke 
OK!
So can I close this PR?

**Comment by Shin-kyoto** [comment]
> Hi @Shin-kyoto Sorry for the late reply.
> 
> As I know there are PRs from @KhalilSelyan #7095 and #8247. He continues my work on this task. And I think this one PR could be closed.

The author @lexavtanke said that  he would like to close this PR, so I will close it. 
@lexavtanke , feel free to re-open it

**Pull request closed without merging**